### PR TITLE
Rework Rx API with FRP concepts

### PR DESCRIPTION
**Pull request by OlivierBlanvillain**
@olafurpg The two middle commits are the interesting bits :)

For reference, [here is](http://package.elm-lang.org/packages/elm-lang/core/3.0.0/Signal) the elm documentation for these functions, and [here is](https://github.com/elm-lang/debug.elm-lang.org/blob/master/examples/Mario.elm) the elm source for the Mario example.

### CODE REVIEW

**Comment by OlivierBlanvillain** [comment]
@olafurpg I did a bunch more reworks, here are the commits that might interest you:

[Rework scala.xml fork](https://github.com/OlivierBlanvillain/monadic-html/pull/39/commits/d3bba1b8de523751690b3f98385853cd0d5d6262
): smaller binaries!

[Modernises Chosen, making heavy use of the cats syntax](https://github.com/OlivierBlanvillain/monadic-html/pull/39/commits/afb8719897812b8bdd3c649850b9b60a1cd43a17): shows how you can write everything using the new functions API.

[Update README](https://github.com/OlivierBlanvillain/monadic-html/pull/39/commits/09262617c2af358a91333ab73f8e816b44f51379): Lots of new documentation!

File: .travis.yml [hunk 1]
```diff
@@ -12,6 +12,7 @@ script:
 - sbt +test
 - sbt 'set scalaJSStage in Global := FullOptStage' test
 - sbt doc
+- grep -Eo "https?://[a-zA-Z0-9./?=_-]+" README.md | while read u; do echo $u; curl -LIs $u | grep "HTTP/1.1 200 OK" || exit 1; done
```
Comment (olafurpg) [inline .travis.yml]: Really nice! I may borrow this one-liner for my own projects :)

File: README.md [hunk 2]
```diff
@@ -93,57 +94,167 @@ When mounting this view, [the implementation](https://github.com/OlivierBlanvill
...
 
 ```scala
 def crazyCanvasStuff(e: dom.html.Canvas): Unit = ...
 
 <canvas mhtml-onmount={ e => crazyCanvasStuff(e) }></canvas>
 ```
 
+## FRP-ish APIs
+
+This section presents the `Rx` API in its entirety. Let's starts with the methods you shouldn't use the impure ones:
```
Comment (olafurpg) [inline README.md]: How about demonstrating the pure api before the impure one?

File: README.md [hunk 3]
```diff
@@ -226,22 +329,27 @@ implicit class SequencingListFFS[A](self: List[Rx[A]]) {
 }
 ```
 
-[Long answer:](https://github.com/typelevel/cats/blob/master/docs/src/main/tut/traverse.md)
+[Long answer:](https://github.com/typelevel/cats/blob/master/docs/src/main/tut/typeclasses/traverse.md)
```
Comment (olafurpg) [inline README.md]: This link http://typelevel.org/cats/typeclasses/traverse.html ?

File: examples/src/main/scala/mhtml/examples/Chosen.scala [hunk 4]
```diff
@@ -29,124 +26,117 @@ object Searcheable {
...
-      val after = toUnderline.substring(index + query.length)
+      val before  = toUnderline.substring(0, index)
+      val after   = toUnderline.substring(index + query.length)
       val matched = toUnderline.substring(index, index + query.length)
-      <span>{before}<u>{matched}</u>{after}</span>
+      <span>{ before }<u>{ matched }</u>{ after }</span>
     }
   }
 
-  def singleSelect[T](getCandidates: String => Rx[Seq[T]],
```
Comment (olafurpg) [inline examples/src/main/scala/mhtml/examples/Chosen.scala]: Thanks for refactoring my crappy code 😆

File: examples/src/main/scala/mhtml/examples/Chosen.scala [hunk 5]
```diff
@@ -29,124 +26,117 @@ object Searcheable {
...
+    val rxFilteredCandidates =
+      (queryEvents |@| candidates).map { case (query, allCandidates) =>
+        allCandidates.filter(Searcheable[T].show(_).toLowerCase.contains(query.toLowerCase))
       }
-      val style = rxFocused.map { focused =>
-        val display = if (focused) "" else "display: none"
-        s"$display"
+
+    val rxIndex: Rx[Int] = (
+        arrowPressedEvents.map(Option(_)) |+|
```
Comment (olafurpg) [inline examples/src/main/scala/mhtml/examples/Chosen.scala]: You need to teach me these `|+|` and `|@|` operators!

File: examples/src/main/scala/mhtml/examples/Chosen.scala [hunk 6]
```diff
@@ -29,124 +26,117 @@ object Searcheable {
...
+        case _ => ()
       }
-      ()
-    }
-    val onblur = { _: Event =>
-      js.timers.setTimeout(300)(rxFocused := false)
-      ()
-    }
+
+    // @olfa: This implementation does not really makes sense as it's a tick
```
Comment (olafurpg) [inline examples/src/main/scala/mhtml/examples/Chosen.scala]: Not sure I follow, what do you mean?

File: monadic-html/src/main/scala/scala/xml/xml.scala [hunk 7]
```diff
@@ -134,68 +138,48 @@ The following types are supported:
...
-  // higher kinded type of kind `(* -> *)` which is a subtype of `Seq[Any]`
-  // (makes no sense), While `C[x] <: Seq[x]` means a higher kinded type of
-  // kind `(* -> *)` which, when instantiated with a type `x`, is a subtype of
-  // `Seq[x]` (what we actually want).
-
-  implicit def seqElementEmbeddable[C[x] <: Seq[x], T <: Node]: XmlElementEmbeddable[C[T]] = instance[C[T]](Group.apply)
-  implicit def rxElementEmbeddable[C[x] <: mhtml.Rx[x], T: XmlElementEmbeddable]: XmlElementEmbeddable[C[T]] = atom[C[T]]
+  type XE[T] = XmlElementEmbeddable[T]
+
+  @inline implicit def nilElementEmbeddable:                              XE[Nil.type]  = null
```
Comment (olafurpg) [inline monadic-html/src/main/scala/scala/xml/xml.scala]: FYI

```
align.tokens = ["=", ":" ]
```

will do this alignment for you in scalafmt ;)

File: monadic-rx-cats/src/main/scala/mhtml/cats.scala [hunk 8]
```diff
@@ -1,8 +1,209 @@
...
+   * val z = run(m) { b =>
+   *   y.cancel
+   *   y = run(f(b)) { b' =>
+   *     x.cancel
+   *     x = run(g(b'))(effect)
+   *   }
+   * }
+   * Cancelable { () => x.cancel; y.cancel; z.cancel }
+   *
+   * Both sides are equivalent. Q.E.D.
```
Comment (olafurpg) [inline monadic-rx-cats/src/main/scala/mhtml/cats.scala]: Mic drop!

I need to read this more detailed later.

File: examples/src/main/scala/mhtml/examples/Chosen.scala (hunk 6 above)
Comment (OlivierBlanvillain) [inline examples/src/main/scala/mhtml/examples/Chosen.scala]: As written, this even triggered 400ms, 800ms, 1200ms ... after the page loaded. Which means, if the text field lost focus at millisecond 700, is would get "unfocused" only 100ms after that.

File: README.md (hunk 3 above)
Comment (OlivierBlanvillain) [inline README.md]: :+1:

File: README.md (hunk 2 above)
Comment (OlivierBlanvillain) [inline README.md]: Good idea!

**Comment by OlivierBlanvillain** [comment]
Thanks for the review!

**Pull request merged**